### ISSUE TITLE
Letter space issue

### ISSUE THREAD

**Issue by vlj** [behdad/harfbuzz#312]
Hi,

I don't think it's an harfbuzz bug, I suspect I'm not using harfbuzz as expected.
I'm trying to implement harfbuzz support in warzone 2100 (an old rts game which use quesoglc atm). The game uses the DejaVuSans font and DejaVuSans-Bold for text rendering on all supported platform, with a character size of 9 (small font), 12 (regular) 16 (medium) and 21 (big).

I followed the implementation of https://github.com/tangrams/harfbuzz-example to write mine which can be found here : https://github.com/vlj/warzone2100/blob/master/lib/ivis_opengl/textdraw.cpp

With bold font disabled,
For reference this is the output using quesoglc :
<img width="577" alt="wz2100glc2" src="https://cloud.githubusercontent.com/assets/6413475/18289265/01e35788-747f-11e6-8780-7c727295e037.PNG">

This is with harfbuzz :
<img width="577" alt="wz2100hb2" src="https://cloud.githubusercontent.com/assets/6413475/18289299/2d0ae020-747f-11e6-9fe8-f8997bf0563b.PNG">

In "Multi Player" the U and the L looks too close. On the other hand the u seems too distant from the uppercase M. Same for the "yer" in "Single Player" : the e seems too close from the r and too distant from the y. Same situation for "Qui" in "Quit Game". It looks like kerning is sometimes applied in the wrong direction (but not always : "Tu" in "Tutorial" looks fine despite the l looking too distant from the a.

I checked with wordpad and the u is not at the same place with DejaVuSans font. A friend on debian also checked on Emacs (not sure it uses harfbuzz but it's likely) and the u is at the same location as it is in wordpad.

I don't know what causes this, if I'm using wrong Freetype2 metrics or not. I'm using harfbuzz package provided by nuget (v0.9.41.1)

**Comment by ebraminio** [comment]
You probably should check [this](https://docs.google.com/document/d/1wpzgGMqXgit6FBVaO76epnnFC_rQPdVKswrDQWyqO1M/edit), likely, some float to int casting is going on on your code.

**Comment by behdad** [comment]
Humm.  Hard to say.  I reviewed https://github.com/tangrams/harfbuzz-example/tree/master/src and that looks alright.  You need to debug this, see what numbers you are getting back and what's expected.

**Comment by vlj** [comment]
I used indeed a texture cache and thus rounded advance to integer position.
Using subpixel positionning (with FT_Set_Transform before loading glyphslot) I get the following which is better (the lower case characters look correctly positionned) :
<img width="577" alt="wz2100hb3" src="https://cloud.githubusercontent.com/assets/6413475/18329563/62528b54-7554-11e6-97ee-fd229e9d9cbf.PNG">

However the leading M in "Multi Player" seems too far from the u. Almost all uppercase followed by lowercase seem too far except in Single player.

**Comment by behdad** [comment]
Is your code public?  You're on the right track, just need more debugging :).

**Comment by vlj** [comment]
Code is here : https://github.com/vlj/warzone2100/blob/master/lib/ivis_opengl/textdraw.cpp
When I set Freetype hinter to FT_LOAD_TARGET_LIGHT the issue is gone :
<img width="577" alt="wz2100hb4" src="https://cloud.githubusercontent.com/assets/6413475/18367938/b69ba348-761d-11e6-8fd0-d81c4ab0b987.PNG">

This surprises me as I didn't tell harfbuzz which hinter I was using. I though harfbuzz wouldn't be influenced by freetype hinting or would relies on metric from the default hinting config.

**Comment by behdad** [comment]
Default load-flags of `hb_ft_font_create` is no-hinting.  So unless you call `hb_ft_font_set_load_flags` to match what flags you use for rendering, you will get mismatching advances and raster.

**Comment by vlj** [comment]
thanks for the explanation !

**Comment by Ident** [comment]
I ran into the exact same issue during the harfbuzz/raqm integration for CEGUI. Btw., coincidentally I also used Deja Vu Sans in the samples, where I tested the rendering result. Setting the load flags as described above, fixed it for me. The resulting positioning was very similar to the Freetype kerning, but between the glyphs where it differed, the harfbuzz positioning was an improvement. So this must have been the correct one.
